### DISCUSSION TITLE
setting up UART (XIOA nrf52840)

### DISCUSSION THREAD

**Discussion by MrTinkerman** [micropython/micropython#14337] [Using MicroPython]
Hi,

I'm trying to talk to a modem using AT commands.
First thing is to get the UART working, I don't get anything back right now.

I'm using a XIOA nrf52840 sense with MicroPython v1.22.2

From the docs:
machine.UART(id,baudrate=115200,bits=8,parity=None,stop=1,tx=None,rx=None)

So I would like to troubleshoot the UART first. Say I use the most basic , default:
```py
modem = UART(0, 115200)
modem.init()
```
How can I now check the parameters bits/parity/stop/tx/rx ?
I want to check the defaults when using the code above.

I also tried initializing with parameters but everything I try gives me various errors, see code below.

Can anyone tell me what I'm doing wrong?

Thanks!

```py
from machine import Pin, UART


# Set up the modem UART
# https://docs.micropython.org/en/latest/library/machine.UART.html
# machine.UART(id,baudrate=115200,bits=8,parity=None,stop=1,tx=None,rx=None)

run = 0

if run==0:
    modem = UART(0, 115200)
    modem.init()
    # No error, but what are the default parameters?

if run==1:
    modem = UART(0,baudrate=115200,bits=8,parity=None,stop=1,tx=None,rx=None)
    # >>> TypeError: extra keyword arguments given

if run==2:
    modem = UART(0, 115200)                         	# init with given baudrate
    modem.init(115200, bits=8, parity=None, stop=1) 	# init with given parameters
    # >>> TypeError: extra positional arguments given

if run==3:
    tx_pin = Pin(43, Pin.OUT)
    rx_pin = Pin(44, Pin.IN)
    modem = UART(0, baudrate=115200, bits=8, parity=None, stop=1, rx=rx_pin, tx=tx_pin)
    # >>> TypeError: extra keyword arguments given

if run==4:
    modem = UART(0, 115200, 8, None, 1, 43, 44)
    modem.init()
    # >>> TypeError: extra positional arguments given

if run==5:
    modem = UART(0, 115200)
    modem.init(bits=8, parity=None, stop=1, rx=44, tx=43)
    # >>> TypeError: extra keyword arguments given

if run==6:
    modem = UART(0, baudrate=115200, tx=Pin(43), rx=Pin(44))
    # >>> TypeError: extra keyword arguments given
```

**Comment by robert-hh** [comment]
The NRF port is pretty different and not really in sync with the other ports.

The UART constructor allows as keyword arguments only `timeout` and `timeout_char`. The baudrate has to be specified as second positional argument. The `init()` method is empty. Print just prints UART(0) as fixed string.

Changing the baudrate requires a hard reset.

**Comment by cwalther** [comment]
In these frustrating cases, it’s usually more helpful to check the [source code](https://github.com/micropython/micropython/blob/d11ca092f75d2652df403dda7448fe1bc8f98cf7/ports/nrf/modules/machine/uart.c#L189) than to try around blindly.

**Comment by MrTinkerman** [comment]
Ok, thanks. 

I tried an ESP32 C3 and I can use:
```py
modem = UART(1, 115200)
modem.init(bits=8, parity=None, stop=1, rx=20, tx=21)
```

No reply from the modem yet though.

Is there a way to 'monitor' the UART to see if it is actually sending data out?

**Reply by robert-hh** [reply]
You can remove `modem.init()` at the nrf port. It is non-functional at the nrf port. All arguments are ignored.
The best method for checking the communication to an external device is using a simple logic analyzer. A $10 device will be sufficient. If you have a spare RP2, you can use that as a logic analyzer, e.g. with this software: https://github.com/dotcypress/ula.git

**Comment by MrTinkerman** [comment]
I used my PPK2 as a logic analyzer and tested both the XIAO nrf52840 and the ESP32C3.
Results below are for the XIAO but I got the same result on the ESP.

So TX goes high and seems to be transmitting when going low. 
Shouldn't the RX (the modems TX) also be high?

Looking at these results I'd say the modem is the issue right?

![XIAOnrf52840_overview](https://github.com/micropython/micropython/assets/16565076/7537f4fa-c747-4a76-9a18-b0156c242aef)

![XIAOnrf52840_detail](https://github.com/micropython/micropython/assets/16565076/190f9023-b78c-4d76-b510-4a1526e37215)

![XIAOnrf52840_tx](https://github.com/micropython/micropython/assets/16565076/e898e265-67dd-40a6-8d02-5a04047cdb99)

**Reply by robert-hh** [reply]
Yes, both TX and RX should be high when idle. I do not really understand your pictures. Only at the last one the bottom track looks like  a TX data transmission. And just to be sure: You connected all of TX, RX and GND, with the modem being powered? And the modem uses 3V3 levels and not V24?.

Another thing you could try is to swap TX and RX at the modem. Sometimes pin are labeled according to a signal they have to be connected to. So a TX pins would then have to be connected to TX at the board.

**Reply by MrTinkerman** [reply]
These are screenshots from Power Profiler, the software that comes with the Power Profiler Kit, PPK2.
It shows power consumption over time (main graph) and up to 8 logic levels below.

So yeah, the RX should be high when idle, why it is not is a mystery at the moment.

Yes everything is connected, power, grounds between boards, TX to RX and I tried swapping too. The logic level is 3V3.
Could it be a power issue? I have a 400mAh 10C lipo powering both boards and I don't think that's the issue.

The topic started out as a question on using the UART but is turning into:
"How to use the Pico Sim7080G with MicroPython".

Maybe I should start a new topic with that title?

**Reply by robert-hh** [reply]
Does the modem something like an enable or cs pin? I saw a SIM7080G breakout documentation a pin called PWRKEY, which must be pulled low for a short time to power up the device.

**Reply by MrTinkerman** [reply]
Yeah, that's were this topic is turning into "How to use the pico SIM7080G with micropython"

Indeed it does have PWR and DTR pins, but I think the use of this board is a topic on its own.
